apps/juxtaposition-ui/src/middleware/webAuth.ts

```diff
@@ -0,0 +1,97 @@
+import { getUserSettings } from '@/api/settings';
+import { config } from '@/config';
+import { accountIsModerator, getUserAccountData, getUserDataFromToken } from '@/util';
+import type { GetUserDataResponse } from '@pretendonetwork/grpc/account/get_user_data_rpc';
+import type { RequestHandler, Request, Response } from 'express';
+
+const cookieDomain = config.http.cookieDomain;
+
+export const webAuth: RequestHandler = async (request, response, next) => {
+	const oauthToken = request.cookies.access_token;
+	request.tokens = { oauthToken };
+
+	// OAuth token auth
+	let pnid: GetUserDataResponse | null = null;
+	// Always auth for guests or write requests
+	if (!request.session.pnid || request.isWrite || !request.cookies.access_token) {
+		const pid = await getUserDataFromToken(request.cookies.access_token).then(tokenUser => tokenUser.pid, _rej => null);
+
+		if (pid !== null) {
+			pnid = await getUserAccountData(pid);
+		}
+
+		// Cache auth result
+		request.session.pnid = pnid;
+	} else {
+		// Use the cache
+		pnid = request.session.pnid;
+	}
```

Comment:
This section could have less nesting, similar to what I suggested in `consoleAuth.ts`